### PR TITLE
Add default property setting to EVP

### PR DESCRIPTION
**Pull request by levitte**
## EVP_set_default_properties(): New function to set global properties

EVP_MD_fetch() can be given a property query string.  However, there
are cases when it won't, for example in implicit fetches.  Therefore,
we also need a way to set a global property query string to be used in
all subsequent fetches.  This also applies to all future algorithm
fetching functions.

## EVP configuration section: add 'default_properties' command

The value of the 'default_properties' command is simply passed to
EVP_set_default_properties().

### CODE REVIEW

**Comment by levitte** [comment]
Note that this depends on #8680 for proper building

File: crypto/evp/evp_cnf.c [hunk 1]
```diff
@@ -37,13 +42,17 @@ static int alg_module_init(CONF_IMODULE *md, const CONF *cnf)
                 return 0;
```
Comment (paulidale) [inline crypto/evp/evp_cnf.c]: missing blank line after `int m;` above.

File: doc/man3/EVP_set_default_properties.pod [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ int EVP_set_default_properties(OPENSSL_CTX *libctx, const char *propq);
+
+=head1 DESCRIPTION
+
+EVP_set_default_properties() set default properties for all future EVP
```
Comment (paulidale) [inline doc/man3/EVP_set_default_properties.pod]: s/set/sets/

File: doc/man3/EVP_set_default_properties.pod [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+algorithm fetches, implicit as well as explicit.
+
+=for comment We should consider having an EVP document in section 7
+that details everything about implicit vs explicit fetches and how
+that relates to properties.
```
Comment (paulidale) [inline doc/man3/EVP_set_default_properties.pod]: s/that relates/they relate/

File: doc/man3/EVP_MD_fetch.pod [hunk 4]
```diff
@@ -39,9 +39,10 @@ algorithm from the default provider.
 
 With explicit fetch an application uses the EVP_MD_fetch() function to obtain
 an algorithm implementation. An implementation with the given name and
-satisfying the search criteria specified in the B<properties> parameter will be
-looked for within the available providers and returned. See L<OSSL_PROVIDER(3)>
-for information about providers.
+satisfying the search criteria specified in the B<properties> parameter
+combined with the default search criteria will be looked for within the
```
Comment (mattcaswell) [inline doc/man3/EVP_MD_fetch.pod]: Perhaps a reference to EVP_set_default_properties would be good here.

File: doc/man5/config.pod [hunk 5]
```diff
@@ -277,9 +277,25 @@ available to the provider.
 This modules has the name B<alg_section> which points to a section containing
```
Comment (paulidale) [inline doc/man5/config.pod]: This modules?

This module reads better.

File: doc/man3/EVP_set_default_properties.pod [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+EVP_set_default_properties stores the properties given with the string
+I<propq> among the EVP data that's been stored in the library context
+given with I<libctx>.
```
Comment (mattcaswell) [inline doc/man3/EVP_set_default_properties.pod]: Explicitly say that libctx may be NULL?

File: doc/man5/config.pod [hunk 7]
```diff
@@ -277,9 +277,25 @@ available to the provider.
...
 algorithm commands.
 
-Currently the only algorithm command supported is B<fips_mode> whose
-value can only be the boolean string B<off>. If B<fips_mode> is set to B<on>,
-an error occurs as this library version is not FIPS capable.
+The supported algorithms commands are:
```
Comment (paulidale) [inline doc/man5/config.pod]: s/algorithms/algorithm/

File: doc/man5/config.pod [hunk 8]
```diff
@@ -277,9 +277,25 @@ available to the provider.
...
+
+=over 4
+
+=item B<default_properties>
+
+The value may be anything that is acceptable as property query string for
```
Comment (mattcaswell) [inline doc/man5/config.pod]: as *a* property query string

File: doc/man5/config.pod [hunk 9]
```diff
@@ -277,9 +277,25 @@ available to the provider.
...
+EVP_set_default_properties().
+
+=item B<fips_mode> (deprecated)
```
Comment (paulidale) [inline doc/man5/config.pod]: ?? Deprecated when it is introduced ??

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_cnf.c [hunk 10]
```diff
@@ -37,13 +42,17 @@ static int alg_module_init(CONF_IMODULE *md, const CONF *cnf)
...
             }
             if (m > 0) {
-                EVPerr(EVP_F_ALG_MODULE_INIT, EVP_R_FIPS_MODE_NOT_SUPPORTED);
-                return 0;
+                EVP_set_default_properties(NULL, "fips=yes");
+            } else {
+                /* TODO(3.0): remove the single propery "fips=yes" */
             }
+        } else if (strcmp(oval->name, "default_properties") == 0) {
+            EVP_set_default_properties(NULL, oval->value);
```
Comment (paulidale) [inline crypto/evp/evp_cnf.c]: Should this merge with the `fips=yes`?

I suggest that they should merge or that we lose the fips= version.

The documentation should indicate that the properties replace any previous ones.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: doc/man5/config.pod (hunk 9 above)
Comment (levitte) [reply doc/man5/config.pod]: This is present in 1.0.2, and is still documented in 1.1.0 and 1.1.1, so not very "introduced" here.
Considering that the whole "fips mode" thing really belongs with the old FOM and that we only preserve the notion for backward compatibility, I do think that deprecation is appropriate.

File: crypto/evp/evp_cnf.c (hunk 10 above)
Comment (levitte) [reply crypto/evp/evp_cnf.c]: Should we have functionality to add a property to an already existing global property query spec?  You might note the comment about removing "fips=yes", just a few lines up...

Comment (levitte) [reply crypto/evp/evp_cnf.c]: And this is a reason for deprecating `fips_mode`.  Its use is disruptive and should be discouraged

File: doc/man5/config.pod (hunk 9 above)
Comment (paulidale) [reply doc/man5/config.pod]: Good point.  I'd really like to remove it though.

Maybe document that setting the default properties in any manner replaces any previously existing default properties.

File: crypto/evp/evp_cnf.c (hunk 10 above)
Comment (paulidale) [reply crypto/evp/evp_cnf.c]: I'd say we don't but do need to document the replacement.

Comment (levitte) [reply crypto/evp/evp_cnf.c]: I've added a bit of documentation on this.

File: doc/man3/EVP_MD_fetch.pod (hunk 4 above)
Comment (levitte) [reply doc/man3/EVP_MD_fetch.pod]: Done

File: doc/man3/EVP_set_default_properties.pod (hunk 6 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: Done

File: doc/man5/config.pod (hunk 8 above)
Comment (levitte) [reply doc/man5/config.pod]: Done

File: crypto/evp/evp_cnf.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/evp_cnf.c]: Okie

File: doc/man3/EVP_set_default_properties.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: Fixed

File: doc/man3/EVP_set_default_properties.pod (hunk 3 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: Fixed

File: doc/man5/config.pod (hunk 5 above)
Comment (levitte) [reply doc/man5/config.pod]: Fixed

File: doc/man5/config.pod (hunk 7 above)
Comment (levitte) [reply doc/man5/config.pod]: Fixed

File: doc/man5/config.pod (hunk 9 above)
Comment (levitte) [reply doc/man5/config.pod]: >  I'd really like to remove it though.

So would I.  Is that permissible or will people get utterly confused?

> Maybe document that setting the default properties in any manner replaces any previously existing default properties.

Good point

File: crypto/evp/evp_cnf.c [hunk 11]
```diff
@@ -37,13 +42,17 @@ static int alg_module_init(CONF_IMODULE *md, const CONF *cnf)
                 return 0;
             }
             if (m > 0) {
-                EVPerr(EVP_F_ALG_MODULE_INIT, EVP_R_FIPS_MODE_NOT_SUPPORTED);
-                return 0;
+                EVP_set_default_properties(NULL, "fips=yes");
+            } else {
+                /* TODO(3.0): remove the single propery "fips=yes" */
```
Comment (paulidale) [inline crypto/evp/evp_cnf.c]: Maybe remove the else branch entirely?

File: doc/man3/EVP_set_default_properties.pod [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ int EVP_set_default_properties(OPENSSL_CTX *libctx, const char *propq);
+
+=head1 DESCRIPTION
+
+EVP_set_default_properties() sets default properties for all future EVP
```
Comment (paulidale) [inline doc/man3/EVP_set_default_properties.pod]: sets **the** default properties

File: doc/man3/EVP_set_default_properties.pod [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+=for comment TODO(3.0) We should consider having an EVP document in
+section 7 that details everything about implicit vs explicit fetches
+and how they relate to properties.
+
+EVP_set_default_properties stores the properties given with the string
+I<propq> among the EVP data that's been stored in the library context
+given with I<libctx> (NULL signifies the default library context).
+
+Any previous default property for the given library context will be
```
Comment (paulidale) [inline doc/man3/EVP_set_default_properties.pod]: _specified_ instead of _given_ perhaps?

File: doc/man5/config.pod [hunk 14]
```diff
@@ -274,12 +274,33 @@ available to the provider.
...
+B<yes>, this is exactly equivalent to:
+
+    default_properties = fips=yes
+
+If the value is B<no>, nothing happens.
+
+=back
+
+These two commands should not be used together, as there is no control
+of how they affect each other.
```
Comment (paulidale) [inline doc/man5/config.pod]: s/of/over/

**Review by paulidale — APPROVED** [review]
a few trivial nits

**Comment by mattcaswell** [comment]
"make update" issues in travis

File: crypto/evp/evp_cnf.c (hunk 11 above)
Comment (levitte) [reply crypto/evp/evp_cnf.c]: Sure

File: doc/man3/EVP_set_default_properties.pod (hunk 12 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: Okie

File: doc/man3/EVP_set_default_properties.pod (hunk 13 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: Sure.

File: doc/man5/config.pod (hunk 14 above)
Comment (levitte) [reply doc/man5/config.pod]: Fixed

**Comment by levitte** [comment]
> "make update" issues in travis

Fixed

**Comment by levitte** [comment]
Now that #8680 is merged, I rebased this PR

File: doc/man3/EVP_set_default_properties.pod [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+ #include <openssl/evp.h>
+
+ int EVP_set_default_properties(OPENSSL_CTX *libctx, const char *propq);
+
+=head1 DESCRIPTION
+
+EVP_set_default_properties() sets the default properties for all
+future EVP algorithm fetches, implicit as well as explicit.
+
+=for comment TODO(3.0) We should consider having an EVP document in
```
Comment (paulidale) [inline doc/man3/EVP_set_default_properties.pod]: Yes!

**Review by paulidale — APPROVED** [review]

File: doc/man3/EVP_set_default_properties.pod (hunk 15 above)
Comment (levitte) [reply doc/man3/EVP_set_default_properties.pod]: > Yes!

I love your enthusiasm :-)

**Comment by levitte** [comment]
Merged.

de3955f66225e42bfae710c50b51c98aa4616ac1 EVP configuration section: add 'default_properties' command
cb92964563a053d5d9c0810912fa6d3ff35c1e16 EVP_set_default_properties(): New function to set global properties

**Pull request closed without merging**